// one-ui/config/example.js
module.exports = {
  columns: [ {
    key: 'username',
    title: 'username',
    width: '20%',
    scaleable: true,
    ellipsis: true,
    sortable: true
  }, {
    key: 'email',
    title: 'email',
    width: '50%',
    scaleable: true,
    ellipsis: true,
    sortable: true
  }, {
    key: 'job',
    title: 'job',
    scaleable: true,
    ellipsis: true,
    sortable: true
  } ],
  peoples: [ {
    username: 'jiangyiqiang',
    email: 'jiangyiqiang@oneapm.com',
    job: 'web developer'
  }, {
    username: 'fuxiaosong',
    email: 'fuxiaosong@oneapm.com',
    job: 'master'
  }, {
    username: 'majing',
    email: 'majing@oneapm.com',
    job: 'web developer'
  }, {
    username: 'leixuewei',
    email: 'leixuewei@oneapm.com',
    job: 'web developer'
  }, {
    username: 'caowen',
    email: 'caowen@oneapm.com',
    job: 'web developer'
  }, {
    username: 'zhangchao',
    email: 'zhangchao@oneapm.com',
    job: 'web developer'
  }, {
    username: 'huanghaiyang',
    email: 'huanghaiyang@oneapm.com',
    job: 'web developer'
  }, {
    username: 'xuyan',
    email: 'xuyan@oneapm.com',
    job: 'web developer'
  } ]
};
