// one-ui/src/components/copy-clipboard/index.js
import $ from 'jquery';
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import ZeroClipboard from 'zeroclipboard';

class CopyClipboard extends Component{
  static defaultProps = {
    text: 'Copy Text',
    staticPath: '/sources/ZeroClipboard.swf'
  };

  constructor(props) {
    super(props);

    this.isCopied = false;

    ZeroClipboard.config({
      swfPath: props.staticPath
    });
  }

  componentDidMount() {
    var element = ReactDOM.findDOMNode(this),
        client = new ZeroClipboard(element);

    this.client = client;

    client.on('ready', () => {
      client.on('beforecopy', () => {
        client.setText(this.props.text);
      });

      client.on('aftercopy', function() {
        if (this.isCopied === true) return;

        var $result = $('<span class="copied" style="position: relative; margin-left: 5px">已复制</span>');

        $(element).append($result);

        setTimeout(() => {
          $result.fadeOut(() => {
            $result.remove();
            this.isCopied = false;
          });
        }, 1000);

        this.isCopied = true;
      });
    });
  }

  componentWillUnmount() {
    this.client.destroy();
  }

  render() {
    return this.props.children || null;
  }
}

export default CopyClipboard;
