// one-ui/build/components/switch-button/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SwitchButton = (_temp = _class = function (_Component) {
  _inherits(SwitchButton, _Component);

  function SwitchButton(props) {
    _classCallCheck(this, SwitchButton);

    var _this = _possibleConstructorReturn(this, (SwitchButton.__proto__ || Object.getPrototypeOf(SwitchButton)).call(this, props));

    _this.state = {
      isChecked: props.checked || false
    };
    return _this;
  }

  _createClass(SwitchButton, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        isChecked: nextProps.checked
      });
    }
  }, {
    key: 'stopPropagation',
    value: function stopPropagation(event) {
      event.stopPropagation();

      if (event.nativeEvent) {
        event.nativeEvent.stopImmediatePropagation();
      }
    }
  }, {
    key: 'toggle',
    value: function toggle(event) {
      if (this.props.disabled) return;

      this.setState({
        isChecked: !this.state.isChecked
      });

      this.props.onChange(this.state.isChecked, event);
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)('trigger', {
            'checked': this.state.isChecked,
            'disabled': this.props.disabled
          }), onClick: this.stopPropagation.bind(this) },
        _react2.default.createElement(
          'label',
          null,
          _react2.default.createElement(
            'span',
            { className: 'trigger-circle' },
            _react2.default.createElement('input', {
              type: 'checkbox',
              checked: this.state.isChecked,
              onChange: this.toggle.bind(this)
            })
          )
        )
      );
    }
  }]);

  return SwitchButton;
}(_react.Component), _class.defaultProps = {
  onChange: function onChange() {},
  checked: false,
  disabled: false
}, _class.propTypes = {
  onChange: _propTypes2.default.func.isRequired,
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool
}, _temp);
exports.default = SwitchButton;
module.exports = exports['default'];
