// one-ui/docs/components/param-table/index.jsx
import $ from 'jquery';
import React from 'react';
import { Table } from '../../../src/components';

const columns = [
  {
    key: 'icon',
    width: 36,
    render() {
      return <i className="icon icon-qidong"></i>;
    }
  },
  {
    key: 'name',
    sortable: true,
    scaleable: true,
    ellipsis: true,
    title: '参数名称'
  },
  {
    key: 'type',
    sortable: true,
    scaleable: true,
    title: '类型',
    format(value) {
      return <code>{ value }</code>;
    }
  },
  {
    key: 'defaults',
    sortable: true,
    scaleable: true,
    title: '默认值'
  },
  {
    key: 'isRequired',
    sortable: true,
    scaleable: true,
    title: '是否必填',
    format(value) {
      return <code>{ `${ Boolean(value) }` }</code>;
    }
  },
  {
    key: 'allow',
    sortable: true,
    scaleable: true,
    title: '可选值',
    format(value = '--') {
      return value;
    }
  },
  {
    key: 'remark',
    sortable: true,
    title: '备注',
    format(value = '--') {
      return value;
    }
  }
];

class ParamTable extends React.Component {
  constructor(props) {
    super(props);

    this.resize = this.resize.bind(this);

    this.width = null;

    $(window).on('resize', this.resize);
  }

  resize() {
    let width = $(window).width() - 250 - 70;

    this.refs.container.style.width = width + 'px';
  }

  componentWillUnmount() {
    $(window).off('resize', this.resize);
  }

  render() {
    return (
      <div className="table-container scroll-x" ref="container">
        <Table columns={ columns } {...this.props} className="param-table" />
      </div>
    );
  }
};

export default ParamTable;
