// one-ui/docs/examples/alert.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import Notify from 'notify';
import ParamTable from '../components/param-table/index.jsx';

class AlertExamples extends React.Component {
  parameters = [{
    name: 'id',
    type: 'String',
    allow: '--',
    defaults: <code>undefined</code>,
    remark: '提示框ID'
  }, {
    name: 'autoClose',
    type: 'Boolean',
    allow: '--',
    defaults: <code>false</code>,
    remark: '是否自动关闭'
  }, {
    name: 'noticeType',
    type: 'String',
    allow: <code>['success', 'info', 'warning', 'error']</code>,
    defaults: <code>undefined</code>,
    remark: '提示框主题'
  }];

  render() {
    return (
      <Page alias="警告框">
        <div className="section">
          <h3>静态示例</h3>
          <div className="example-area">
            <div className="alert alert-danger">
              <p>警告框警告框警告框警告框警告框警告框警告框警告框警告框。。。</p>
              <button type="button" className="close" data-dismiss="alert">
                <i className="icon icon-close"></i>
              </button>
            </div>

            <div className="alert alert-primary">
              <p>警告框警告框警告框警告框警告框警告框警告框警告框警告框。。。</p>
              <button type="button" className="close" data-dismiss="alert">
                <i className="icon icon-close"></i>
              </button>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="alert alert-danger">
  <p>警告框警告框警告框警告框警告框警告框警告框警告框警告框。。。</p>
  <button type="button" class="close" data-dismiss="alert">
    <i class="icon icon-close"></i>
  </button>
</div>

<div class="alert alert-primary">
  <p>警告框警告框警告框警告框警告框警告框警告框警告框警告框。。。</p>
  <button type="button" class="close" data-dismiss="alert">
    <i class="icon icon-close"></i>
  </button>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>使用示例</h3>
          <div className="example-area">
            <Notify noticeType="error">警告框警告框警告框警告框警告框警告框警告框警告框警告框。。。</Notify>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<Notify noticeType="error">警告框警告框警告框警告框警告框警告框警告框警告框警告框。。。</Notify>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <div className="example-area">
            <ParamTable data={ this.parameters } />
          </div>
        </div>
      </Page>
    );
  }
}

export default AlertExamples;
