// one-ui/docs/examples/dropdown.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import Dropdown from 'dropdown';
import ParamTable from '../components/param-table/index.jsx';

class DropdownExamples extends React.Component {
  parameters = [
    {
      name: 'autoHide',
      type: 'Boolean',
      defaults: <code>false</code>,
      remark: '是否自动关闭'
    }, {
      name: 'labelClassName',
      type: 'Any',
      defaults: <code>undefined</code>,
      remark: '按钮样式类'
    }, {
      name: 'label',
      type: 'Any',
      defaults: <code>undefined</code>,
      remark: '按钮内容'
    }, {
      name: 'pullRight',
      type: 'Boolean',
      defaults: <code>false</code>,
      remark: '对齐方式'
    }, {
      name: 'menuStyle',
      type: 'Object',
      defaults: <code>{'{}'}</code>,
      remark: '下拉菜单样式'
    }
  ];

  render() {
    return (
      <Page alias="下拉菜单">
        <div className="section">
          <h3>样式示例</h3>
          <p>将下拉菜单触发器和下拉菜单都包裹在 <code>.dropdown</code> 里，或者另一个声明了 <code>position: relative;</code> 的元素。然后加入组成菜单的 HTML 代码。</p>
          <div className="example-area pure-g">
            <div className="pure-u-1-2">
              <div className="dropdown open">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  选择器
                </button>
                <div className="dropdown-menu" style={{ position: 'relative' }}>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>

            <div className="pure-u-1-2">
              <div className="dropdown open">
                <button className="btn dropdown-toggle btn-shadow">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  选择器
                </button>
                <div className="dropdown-menu" style={{ position: 'relative' }}>
                  <span className="arrow"></span>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="dropdown">
  <button class="btn dropdown-toggle pull-right">
    选择器
    <span class="icon" style="margin-left: 10px;">&#xe64b;</span>
  </button>
  <div class="dropdown-menu pull-right">
    <div class="dropdown-filter">
      <i class="icon">&#xe647;</i>
      <input type="text" class="form-control" placeholder="搜索事件" />
    </div>
    <ul>
      <li class="item"><a href="javascript:;">全部实例</a></li>
      <li class="item"><a href="javascript:;">下拉选项</a></li>
      <li class="item"><a href="javascript:;">下拉选项</a></li>
      <li class="item"><a href="javascript:;">下拉选项</a></li>
    </ul>
  </div>
</div>

<div class="dropdown">
  <button class="btn dropdown-toggle btn-shadow">
    <span class="caret icon" style="margin-left: 10px;">&#xe64b;</span>
    选择器
  </button>
  <div class="dropdown-menu">
    <span class="arrow"></span>
    <div class="dropdown-filter">
      <i class="icon">&#xe647;</i>
      <input type="text" class="form-control" placeholder="搜索事件" />
    </div>
    <ul>
      <li class="item"><a href="javascript:;">全部实例</a></li>
      <li class="item"><a href="javascript:;">下拉选项</a></li>
      <li class="item"><a href="javascript:;">下拉选项</a></li>
      <li class="item"><a href="javascript:;">下拉选项</a></li>
    </ul>
  </div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>对齐</h3>
          <p>默认情况下，下拉菜单自动沿着父元素的上沿和左侧被定位为 100% 宽度。 为 <code>.dropdown-menu</code> 添加 <code>.dropdown-menu-right</code> 类可以让菜单右对齐。注意：当 .dropdown-menu 的宽度小于或等于 100% 的时候是看不到对齐效果的！</p>
          <p>如果需要选择对齐方式，请使用 <code>.dropdown-menu-left</code> 或是 <code>.dropdown-menu-right</code> 样式类，不要使用 .pull-right 样式类。</p>
          <div className="example-area pure-g">
            <div className="pure-u-1-2">
              <div className="dropdown open">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  选择器
                </button>
                <div className="dropdown-menu" style={{ position: 'relative' }}>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>
            <div className="pure-u-1-2">
              <div className="dropdown open">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  选择器
                </button>
                <div className="dropdown-menu dropdown-menu-right">
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例全部实例全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div className="dropdown open">
  <button className="btn dropdown-toggle">
    <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
    选择器
  </button>
  <div className="dropdown-menu" style={{ position: 'relative' }}>
    <div className="dropdown-filter">
      <i className="icon">&#xe647;</i>
      <input type="text" className="form-control" placeholder="搜索事件" />
    </div>
    <ul>
      <li className="item"><a href="javascript:;">全部实例</a></li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
    </ul>
  </div>
</div>

<div className="dropdown open">
  <button className="btn dropdown-toggle">
    <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
    选择器
  </button>
  <div className="dropdown-menu dropdown-menu-right">                         <!-- 使用 .dropdown-menu-right 类设置对齐方式 -->
    <div className="dropdown-filter">
      <i className="icon">&#xe647;</i>
      <input type="text" className="form-control" placeholder="搜索事件" />
    </div>
    <ul>
      <li className="item"><a href="javascript:;">全部实例全部实例全部实例</a></li>
      <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
      <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
      <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
    </ul>
  </div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>带箭头</h3>
          <p>当然你也可以选择带小箭头的下拉框，在 <code>.dropdown</code> 元素上加上 <code>.dropdown-enable-arrow</code> 样式类并在放置小箭头元素即可。</p>
          <div className="example-area pure-g">
            <div className="pure-u-1-2">
              <div className="dropdown open dropdown-enable-arrow">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  选择器
                </button>
                <div className="dropdown-menu" style={{ position: 'relative' }}>
                  <i className="arrow"></i>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>
            <div className="pure-u-1-2">
              <div className="dropdown open dropdown-enable-arrow">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  选择器
                </button>
                <div className="dropdown-menu dropdown-menu-right">
                  <i className="arrow"></i>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例全部实例全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
                    <li className="item"><a href="javascript:;">下拉选项下拉选项下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div className="dropdown open dropdown-enable-arrow">                 <!-- 添加 .dropdown-enable-arrow 开启小箭头 -->
  ...
  <div className="dropdown-menu" style={{ position: 'relative' }}>
    <i class="arrow"></i>                                             <!-- 添加 .arrow 节点 -->
    ...
  </div>
</div>

<div className="dropdown open dropdown-enable-arrow">                 <!-- 添加 .dropdown-enable-arrow 开启小箭头 -->
  ...
  <div className="dropdown-menu dropdown-menu-right">
    <i class="arrow"></i>                                             <!-- 添加 .arrow 节点 -->
    ...
  </div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>多级菜单</h3>
          <p>如果你需要多级菜单，请在对应的 <code>.item</code> 选项中加入 <code>.menu-parent</code> 类并在该节点下添加对应的子菜单结构。</p>
          <div className="example-area pure-g">
            <div className="pure-u-1-2">
              <div className="dropdown open">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  多层级的选择器
                </button>
                <div className="dropdown-menu" style={{ position: 'relative' }}>
                  <span className="arrow"></span>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item menu-parent">
                      <a href="javascript:;">二级菜单</a>
                      <ul className="dropdown-menu">
                        <li className="item"><a href="javascript:;">子菜单</a></li>
                        <li className="item"><a href="javascript:;">子菜单</a></li>
                        <li className="item"><a href="javascript:;">子菜单</a></li>
                      </ul>
                    </li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>

            <div className="pure-u-1-2">
              <div className="dropdown open">
                <button className="btn dropdown-toggle">
                  <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                  多层级的选择器
                </button>
                <div className="dropdown-menu" style={{ position: 'relative' }}>
                  <span className="arrow"></span>
                  <div className="dropdown-filter">
                    <i className="icon">&#xe647;</i>
                    <input type="text" className="form-control" placeholder="搜索事件" />
                  </div>
                  <ul>
                    <li className="item"><a href="javascript:;">全部实例</a></li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                    <li className="item menu-parent">
                      <a href="javascript:;">二级菜单</a>
                      <ul className="dropdown-menu align-left">
                        <li className="item"><a href="javascript:;">子菜单</a></li>
                        <li className="item"><a href="javascript:;">子菜单</a></li>
                        <li className="item"><a href="javascript:;">子菜单</a></li>
                      </ul>
                    </li>
                    <li className="item"><a href="javascript:;">下拉选项</a></li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre>
              <code>
{`<div className="dropdown open">
  ...
  <div className="dropdown-menu" style={{ position: 'relative' }}>
    ...
    <ul>
      <li className="item"><a href="javascript:;">全部实例</a></li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
      <li className="item menu-parent">                               <!-- 二级菜单 -->
        <a href="javascript:;">二级菜单</a>
        <ul className="dropdown-menu">                                <!-- 二级菜单选项，同一级菜单结构 -->
          <li className="item"><a href="javascript:;">子菜单</a></li>
          <li className="item"><a href="javascript:;">子菜单</a></li>
          <li className="item"><a href="javascript:;">子菜单</a></li>
        </ul>
      </li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
    </ul>
  </div>
</div>

<div className="dropdown open">
  ...
  <div className="dropdown-menu" style={{ position: 'relative' }}>
    ...
    <ul>
      <li className="item"><a href="javascript:;">全部实例</a></li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
      <li className="item menu-parent">                               <!-- 二级菜单 -->
        <a href="javascript:;">二级菜单</a>
        <ul className="dropdown-menu align-left">                     <!-- 二级菜单选项，同一级菜单结构 -->
          <li className="item"><a href="javascript:;">子菜单</a></li>
          <li className="item"><a href="javascript:;">子菜单</a></li>
          <li className="item"><a href="javascript:;">子菜单</a></li>
        </ul>
      </li>
      <li className="item"><a href="javascript:;">下拉选项</a></li>
    </ul>
  </div>
</div>`}
              </code>
            </pre>
          </div>
        </div>

        <div className="section">
          <h3>小尺寸</h3>
          <div className="example-area">
            <div className="dropdown open">
              <button className="btn dropdown-toggle dropdown-toggle-mini">
                <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
                多层级的选择器
              </button>
              <div className="dropdown-menu" style={{ position: 'relative' }}>
                <span className="arrow"></span>
                <div className="dropdown-filter">
                  <i className="icon">&#xe647;</i>
                  <input type="text" className="form-control" placeholder="搜索事件" />
                </div>
                <ul>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">下拉选项</a></li>
                  <li className="item menu-parent">
                    <a href="javascript:;">二级菜单</a>
                    <ul className="dropdown-menu">
                      <li className="item"><a href="javascript:;">子菜单</a></li>
                      <li className="item"><a href="javascript:;">子菜单</a></li>
                      <li className="item"><a href="javascript:;">子菜单</a></li>
                    </ul>
                  </li>
                  <li className="item"><a href="javascript:;">下拉选项</a></li>
                </ul>
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre>
              <code>
{`<div className="dropdown open">
  <button className="btn dropdown-toggle dropdown-toggle-mini">                   <!-- 添加 .dropdown-toggle-mini 类设置为小尺寸 -->
    <span className="caret icon" style={{ marginLeft: 10 }}>&#xe64b;</span>
    多层级的选择器
  </button>
  ...
</div>`}
              </code>
            </pre>
          </div>
        </div>

        <div className="section">
          <h3>使用示例</h3>
          <div className="example-area pure-g">
            <div className="pure-u-1-3">
              <Dropdown label="请选择查询时间" labelClassName="dropdown-toggle-mini">
                <span className="arrow"></span>
                <div className="dropdown-filter">
                  <i className="icon">&#xe647;</i>
                  <input type="text" className="form-control" placeholder="搜索事件" />
                </div>
                <ul>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                </ul>
              </Dropdown>
            </div>

            <div className="pure-u-1-3">
              <Dropdown label="请选择查询时间">
                <span className="arrow"></span>
                <div className="dropdown-filter">
                  <i className="icon">&#xe647;</i>
                  <input type="text" className="form-control" placeholder="搜索事件" />
                </div>
                <ul>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                </ul>
              </Dropdown>
            </div>

            <div className="pure-u-1-3">
              <Dropdown label="请选择查询时间" className="dropdown-enable-arrow" labelClassName="btn-shadow" pullRight>
                <span className="arrow"></span>
                <div className="dropdown-filter">
                  <i className="icon">&#xe647;</i>
                  <input type="text" className="form-control" placeholder="搜索事件" />
                </div>
                <ul>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                  <li className="item"><a href="javascript:;">全部实例</a></li>
                </ul>
              </Dropdown>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<Dropdown label="请选择查询时间" labelClassName="dropdown-toggle-mini">
  <span className="arrow"></span>
  <div className="dropdown-filter">
    <i className="icon">&#xe647;</i>
    <input type="text" className="form-control" placeholder="搜索事件" />
  </div>
  <ul>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
  </ul>
</Dropdown>

<Dropdown label="请选择查询时间">
  <span className="arrow"></span>
  <div className="dropdown-filter">
    <i className="icon">&#xe647;</i>
    <input type="text" className="form-control" placeholder="搜索事件" />
  </div>
  <ul>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
  </ul>
</Dropdown>

<Dropdown label="请选择查询时间" className="dropdown-enable-arrow" labelClassName="btn-shadow" pullRight>
  <span className="arrow"></span>
  <div className="dropdown-filter">
    <i className="icon">&#xe647;</i>
    <input type="text" className="form-control" placeholder="搜索事件" />
  </div>
  <ul>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
    <li className="item"><a href="javascript:;">全部实例</a></li>
  </ul>
</Dropdown>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <ParamTable data={ this.parameters } />
        </div>
      </Page>
    );
  }
}

export default DropdownExamples;
