// one-ui/docs/examples/layout.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import { Layout } from '../../src/components';
import ParamTable from '../components/param-table/index.jsx';

class LayoutExamples extends React.Component {
  parameters = [{
    name: 'type',
    type: 'String',
    defaults: <code>flex</code>,
    allow: <code>['flex', 'grid']</code>,
    remark: '布局类型'
  }, {
    name: 'grid',
    type: 'Number | String',
    defaults: <code>1</code>,
    remark: '占比'
  }, {
    name: 'height',
    type: 'Number | String',
    defaults: <code>undefined</code>,
    remark: '高'
  }, {
    name: 'direction',
    type: 'col',
    defaults: <code>col</code>,
    allow: <code>['col', 'row']</code>,
    remark: '仅限于 flex 布局时使用',
    remark: '流向'
  }, {
    name: 'alignSelf',
    type: 'String',
    defaults: <code>undefined</code>,
    allow: <code>['top', 'center', 'bottom']</code>,
    remark: '仅限于 flex 布局时使用'
  }, {
    name: 'horizontal',
    type: 'String',
    allow: <code>['left', 'center', 'right', 'around', 'between']</code>,
    defaults: <code>undefined</code>,
    remark: '仅限于 flex 布局时使用'
  }, {
    name: 'vertical',
    type: 'String',
    defaults: <code>undefined</code>,
    allow: <code>['top', 'middle', 'bottom']</code>,
    remark: '仅限于 flex 布局时使用'
  }, {
    name: 'attrs',
    type: 'Object',
    defaults: <code>undefined</code>,
    remark: '附加属性'
  }];

  render() {
    return (
      <Page alias="布局">
        <div className="section">
          <h3>伸缩布局</h3>
          <div className="example-area flexible-examples">
            <Layout direction="row" grid={ false }>
              <Layout direction={ null }>
                flexbox
              </Layout>
              <Layout direction={ null }>
                flexbox
              </Layout>
              <Layout direction={ null }>
                flexbox
              </Layout>
              <Layout direction={ null }>
                flexbox
              </Layout>
            </Layout>

            <Layout grid={ false } height={100} vertical="top" className="h-50">
              align-items: top
            </Layout>

            <Layout grid={ false } height={100} vertical="middle" className="h-50">
              align-items: middle
            </Layout>

            <Layout grid={ false } height={100} vertical="bottom" className="h-50">
              align-items: bottom
            </Layout>

            <Layout grid={ false } height={100} direction="row" horizontal="left">
              <span className="flex-text-demo">justify-content: left</span>
            </Layout>

            <Layout grid={ false } height={100} direction="row" horizontal="center">
              <span className="flex-text-demo">justify-content: center</span>
            </Layout>

            <Layout grid={ false } height={100} direction="row" horizontal="right">
              <span className="flex-text-demo">justify-content: right</span>
            </Layout>

            <Layout grid={ false } height={100} direction="row" horizontal="around">
              <span className="flex-text-demo">justify-content: around</span>
              <span className="flex-text-demo">justify-content: around</span>
            </Layout>

            <Layout grid={ false } height={100} direction="row" horizontal="between">
              <span className="flex-text-demo">justify-content: between</span>
              <span className="flex-text-demo">justify-content: between</span>
            </Layout>
          </div>
          <div className="source-code-area">
            <pre><code>
{`<Layout direction="row" grid={ false }>
  <Layout direction={ null }>
    flexbox
  </Layout>
  <Layout direction={ null }>
    flexbox
  </Layout>
  <Layout direction={ null }>
    flexbox
  </Layout>
  <Layout direction={ null }>
    flexbox
  </Layout>
</Layout>

<Layout grid={ false } vertical="top" className="h-50">
  align-items: top
</Layout>

<Layout grid={ false } vertical="middle" className="h-50">
  align-items: middle
</Layout>

<Layout grid={ false } vertical="bottom" className="h-50">
  align-items: bottom
</Layout>

<Layout grid={ false } direction="row" horizontal="left">
  <span className="flex-text-demo">justify-content: left</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="center">
  <span className="flex-text-demo">justify-content: center</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="right">
  <span className="flex-text-demo">justify-content: right</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="around">
  <span className="flex-text-demo">justify-content: around</span>
  <span className="flex-text-demo">justify-content: around</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="between">
  <span className="flex-text-demo">justify-content: between</span>
  <span className="flex-text-demo">justify-content: between</span>
</Layout>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>流式布局</h3>
          <div className="example-area">
            <div className="pure-g grid-unit">
              <Layout type="grid" height={24} grid="1">
                pure-u-1
              </Layout>
            </div>
            <div className="pure-g grid-unit">
              <Layout type="grid" height={24} grid="1-2">
                pure-u-1-2
              </Layout>
            </div>
            <div className="pure-g grid-unit">
              <Layout type="grid" height={24} grid="1-3">
                pure-u-1-3
              </Layout>
            </div>
            <div className="pure-g grid-unit">
              <Layout type="grid" height={24} grid="1-4">
                pure-u-1-4
              </Layout>
            </div>
            <div className="pure-g grid-unit">
              <Layout type="grid" height={24} grid="1-5">
                pure-u-1-5
              </Layout>
            </div>
          </div>
        </div>
        <div className="source-code-area">
          <pre>
            <code>
{`<div className="pure-g">
  <Layout type="grid" grid="1">
    pure-u-1
  </Layout>
</div>

<div className="pure-g">
  <Layout type="grid" grid="1-2">
    pure-u-1-2
  </Layout>
</div>

<div className="pure-g">
  <Layout type="grid" grid="1-3">
    pure-u-1-3
  </Layout>
</div>

<div className="pure-g">
  <Layout type="grid" grid="1-4">
    pure-u-1-4
  </Layout>
</div>

<div className="pure-g">
  <Layout type="grid" grid="1-5">
    pure-u-1-5
  </Layout>
</div>`}
            </code>
          </pre>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <ParamTable data={ this.parameters } />
        </div>
      </Page>
    );
  }
}

export default LayoutExamples;
