// one-ui/docs/examples/modal.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import { Modal } from '../../src/components';
import ParamTable from '../components/param-table/index.jsx';

class ModalExamples extends React.Component {
  parameters = [{
    name: 'modal',
    type: 'Boolean',
    defaults: <code>true</code>,
    remark: '是否显示遮罩层'
  }, {
    name: 'draggable',
    type: 'Boolean',
    defaults: <code>true</code>,
    remark: '是否可拖拽'
  }, {
    name: 'resizable',
    type: 'Boolean',
    defaults: <code>false</code>,
    remark: '是否可缩放'
  }, {
    name: 'className',
    type: 'Any',
    defaults: <code>null</code>,
    remark: '最外层元素样式'
  }, {
    name: 'headerClassName',
    type: 'Any',
    defaults: <code>null</code>,
    remark: '头部样式（主要用于处理头部带选项卡的情况）'
  }, {
    name: 'defaultSize',
    type: 'Object',
    defaults: <code>{ `{width: 'auto', height: 'auto'}` }</code>,
    remark: '默认尺寸'
  }, {
    name: 'defaultPosition',
    type: 'Object',
    defaults: <code>{ `{ top: 30, right: 30, bottom: 30, left: 30}` }</code>,
    remark: '默认位置'
  }, {
    name: 'onClose',
    type: 'Function',
    defaults: <code>{ '() => {}' }</code>,
    remark: '关闭模态框时触发的回调'
  }];

  handleMountModal() {
    Modal.mount({
      title: '测试模态框',
      draggable: false
    });
  }

  render() {
    return (
      <Page alias="模态框">
        <div className="section">
          <h3>静态示例</h3>
          <p>以下模态框包含了模态框的头、体和一组放置于底部的按钮。</p>
          <div className="example-area">
            <div className="modal">
              <div className="modal-content">
                <div className="modal-header">
                  <span className="modal-title">modal title</span>
                  <button className="icon close">&#xe64a;</button>
                </div>
                <div className="modal-body">
                  <div className="table-header">
                    <table className="table table-fixed">
                      <thead>
                        <tr>
                          <th>username</th>
                          <th>email</th>
                          <th>company</th>
                          <th>action</th>
                        </tr>
                      </thead>
                    </table>
                  </div>
                  <div className="table-body" style={{ height: 200, overflow: 'auto' }}>
                    <table className="table table-hover table-striped table-fixed">
                      <tbody>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                        <tr>
                          <td>artisan</td>
                          <td>artisan@126.com</td>
                          <td>oneapm</td>
                          <td>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-trash"></i></a>
                            <a href="javascript:;" className="text-muted"><i className="icon icon-edit"></i></a>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <div className="modal-footer">
                  <button className="btn">取消</button>
                  <button className="btn btn-primary">保存</button>
                </div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="modal">
  <div class="modal-content">
    <div class="modal-header">
      <span class="modal-title">modal title</span>
      <button class="icon close">&#xe64a;</button>
    </div>
    <div class="modal-body">
      ...
    </div>
    <div class="modal-footer">
      <button className="btn">取消</button>
      <button class="btn btn-primary">保存</button>
    </div>
  </div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>模态框（头部附带选项卡）</h3>
          <p>在 <code>.modal-header</code> 上添加 <code>.modal-tabs</code> 并加入选项卡的代码可实现头部带选项卡的模态框。</p>
          <div className="example-area">
            <div className="modal">
              <div className="modal-content">
                <div className="modal-header modal-tabs">
                  <nav className="nav-tabs">
                    <a href="javascript:;" className="tabs-item">选项一</a>
                    <a href="javascript:;" className="tabs-item active">选项二<span className="icon close">&#xe64a;</span></a>
                    <a href="javascript:;" className="tabs-item">选项三<span className="icon close">&#xe64a;</span></a>
                  </nav>
                  <button className="icon close">&#xe64a;</button>
                </div>
                <div className="modal-body">
                  <nav className="nav-tabs">
                    <li className="tabs-item"><a href="javascript:;">选项一</a></li>
                    <li className="tabs-item active"><a href="javascript:;">选项二</a></li>
                    <li className="tabs-item"><a href="javascript:;">选项三</a></li>
                  </nav>
                  <div className="tab-content" style={{ marginBottom: 0 }}>
                    content...
                  </div>
                </div>
                <div className="modal-footer">
                  <button className="btn btn-primary">保存</button>
                </div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="modal">
  <div class="modal-content">
    <div class="modal-header modal-tabs">
      <nav class="nav-tabs">
        <a href="javascript:;" class="tabs-item">选项一</a>
        <a href="javascript:;" class="tabs-item active">选项二<span class="icon close">&#xe64a;</span></a>
        <a href="javascript:;" class="tabs-item">选项三<span class="icon close">&#xe64a;</span></a>
      </nav>
      <button class="icon close">&#xe64a;</button>
    </div>
    <div class="modal-body">
      <nav class="nav-tabs">
        <li class="tabs-item"><a href="javascript:;">选项一</a></li>
        <li class="tabs-item active"><a href="javascript:;">选项二</a></li>
        <li class="tabs-item"><a href="javascript:;">选项三</a></li>
      </nav>
      <div class="tab-content" style="margin-bottom: 0">
        content...
      </div>
    </div>
    <div class="modal-footer">
      <button class="btn btn-primary">保存</button>
    </div>
  </div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>动态示例</h3>
          <div className="example-area">
            <a href="javascript:;" className="btn btn-primary" onClick={ this.handleMountModal.bind(this) }>点我点我！</a>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <div className="example-area">
            <ParamTable data={ this.parameters } />
          </div>
        </div>
      </Page>
    );
  }
}

export default ModalExamples;
