// one-ui/docs/examples/nav-tab.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import NavTab from 'nav-tab';
import ParamTable from '../components/param-table/index.jsx';

class NavTabExamples extends React.Component {
  parameters = [{
    name: 'className',
    type: 'Any',
    defaults: <code>undefined</code>,
    remark: '提示框ID'
  }, {
    name: 'active',
    type: 'Number',
    defaults: <code>0</code>,
    remark: '激活选项的索引值'
  }, {
    name: 'theme',
    type: 'String',
    allow: <code>['line', 'out']</code>,
    defaults: <code>undefined</code>,
    remark: '选项卡主题'
  }, {
    name: 'rights',
    type: 'Array',
    defaults: <code>[]</code>,
    remark: '右侧工具栏'
  }, {
    name: <span><code>NavTab.Panel</code> Panel.title</span>,
    type: 'String',
    defaults: <code>''</code>,
    remark: '选项卡名称'
  }, {
    name: <span><code>NavTab.Panel</code> Panel.href</span>,
    type: 'String',
    defaults: <code>null</code>,
    remark: '选项卡链接'
  }, {
    name: <span><code>NavTab.Panel</code> Panel.handleClick</span>,
    type: 'Function',
    defaults: <code>function () {}</code>,
    remark: '点击回调函数'
  }, {
    name: <span><code>NavTab.Panel</code> Panel.contentheight</span>,
    type: 'Number | String',
    defaults: <code>auto</code>,
    remark: '内容区域高度'
  }];

  render() {
    return (
      <Page alias="选项卡">
        <div className="section">
          <h3>默认样式</h3>
          <p>默认样式的选项卡。</p>
          <div className="example-area">
            <div className="nav-tabs-example">
              <nav className="nav-tabs">
                <li className="tabs-item"><a href="javascript:;">选项一</a></li>
                <li className="tabs-item"><a href="javascript:;">选项二</a></li>
                <li className="tabs-item active"><a href="javascript:;">选项三</a></li>
              </nav>
              <div className="tab-content">
                选项卡内容
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
  {`<nav className="nav-tabs">
  <li className="tabs-item"><a href="javascript:;">选项一</a></li>
  <li className="tabs-item"><a href="javascript:;">选项二</a></li>
  <li className="tabs-item active"><a href="javascript:;">选项三</a></li>
</nav>
<div className="tab-content">
  选项卡内容
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>带下划线</h3>
          <p>添加 <code>.tabs-line</code> 可以实现带下划线的选项卡。</p>
          <div className="example-area">
            <div className="nav-tabs-example">
              <nav className="nav-tabs tabs-line">
                <li className="tabs-item active"><a href="javascript:;">选项一</a></li>
                <li className="tabs-item"><a href="javascript:;">选项二</a></li>
                <li className="tabs-item"><a href="javascript:;">选项三</a></li>
              </nav>
              <div className="tab-content">
                选项卡内容
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
  {`<nav className="nav-tabs tabs-line">
  <li className="tabs-item active"><a href="javascript:;">选项一</a></li>
  <li className="tabs-item"><a href="javascript:;">选项二</a></li>
  <li className="tabs-item"><a href="javascript:;">选项三</a></li>
</nav>
<div className="tab-content">
  选项卡内容
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>蓝色背景</h3>
          <p>添加 <code>.tabs-out</code> 可以实现带背景色的选项卡。</p>
          <div className="example-area">
            <div className="nav-tabs-example">
              <nav className="nav-tabs tabs-out">
                <li className="tabs-item"><a href="javascript:;">选项一</a></li>
                <li className="tabs-item active"><a href="javascript:;">选项二</a></li>
                <li className="tabs-item"><a href="javascript:;">选项三</a></li>
              </nav>
              <div className="tab-content">
                选项卡内容
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
  {`<nav className="nav-tabs tabs-out">
  <li className="tabs-item"><a href="javascript:;">选项一</a></li>
  <li className="tabs-item active"><a href="javascript:;">选项二</a></li>
  <li className="tabs-item"><a href="javascript:;">选项三</a></li>
</nav>
<div className="tab-content">
  选项卡内容
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>二级导航</h3>
          <p>二级导航与其他类型选项卡的区别在于：字体大小为 <code>1.4rem</code>，高度为 <code>45px</code>。你可以添加 <code>.nav-tabs-navbar</code> 类开启这个效果。</p>
          <div className="example-area">
            <div className="nav-tabs-example">
              <nav className="nav-tabs tabs-line nav-tabs-navbar">
                <li className="tabs-item"><a href="javascript:;">选项一</a></li>
                <li className="tabs-item"><a href="javascript:;">选项二</a></li>
                <li className="tabs-item active"><a href="javascript:;">选项三</a></li>
              </nav>
              <div className="tab-content">
                选项卡内容
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
  {`<nav className="nav-tabs">
  <li className="tabs-item"><a href="javascript:;">选项一</a></li>
  <li className="tabs-item"><a href="javascript:;">选项二</a></li>
  <li className="tabs-item active"><a href="javascript:;">选项三</a></li>
</nav>
<div className="tab-content">
  选项卡内容
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>嵌套演示</h3>
          <p>以下是一个选项卡嵌套使用的效果，需要注意的是：最后一级选项卡的高度为 <code>30px</code>，其余两级的为 <code>35px</code></p>
          <div className="example-area">
            <div className="nav-tabs-example">
              <nav className="nav-tabs tabs-out">
                <li className="tabs-item active"><a href="javascript:;">选项一级</a></li>
                <li className="tabs-item"><a href="javascript:;">选项一级</a></li>
                <li className="tabs-item"><a href="javascript:;">选项一级</a></li>
              </nav>
              <div className="tab-content" style={{padding: 0}}>
                <nav className="nav-tabs tabs-line">
                  <li className="tabs-item active"><a href="javascript:;">选项二级</a></li>
                  <li className="tabs-item"><a href="javascript:;">选项二级</a></li>
                  <li className="tabs-item"><a href="javascript:;">选项二级</a></li>
                </nav>
                <div className="tab-content tabs-inner" style={{padding: 0}}>
                  <nav className="nav-tabs">
                    <li className="tabs-item active"><a href="javascript:;">选项三级</a></li>
                    <li className="tabs-item"><a href="javascript:;">选项三级</a></li>
                    <li className="tabs-item"><a href="javascript:;">选项三级</a></li>
                  </nav>
                  <div className="tab-content">
                    选项卡内容
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
  {`<div className="nav-tabs-example">
  <nav className="nav-tabs tabs-out">
    <li className="tabs-item active"><a href="javascript:;">选项一级</a></li>
    <li className="tabs-item"><a href="javascript:;">选项一级</a></li>
    <li className="tabs-item"><a href="javascript:;">选项一级</a></li>
  </nav>
  <div className="tab-content" style={{padding: 0}}>
    <nav className="nav-tabs tabs-line">
      <li className="tabs-item active"><a href="javascript:;">选项二级</a></li>
      <li className="tabs-item"><a href="javascript:;">选项二级</a></li>
      <li className="tabs-item"><a href="javascript:;">选项二级</a></li>
    </nav>
    <div className="tab-content tabs-inner" style={{padding: 0}}>
      <nav className="nav-tabs">
        <li className="tabs-item active"><a href="javascript:;">选项三级</a></li>
        <li className="tabs-item"><a href="javascript:;">选项三级</a></li>
        <li className="tabs-item"><a href="javascript:;">选项三级</a></li>
      </nav>
      <div className="tab-content">
        选项卡内容
      </div>
    </div>
  </div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>动态示例</h3>
          <div className="example-area">
            <div className="nav-tabs-example">
              <NavTab theme="line">
                <NavTab.Panel key="1" title="选项一" children="选项卡内容一" />
                <NavTab.Panel key="2" title="选项二" children="选项卡内容二" />
                <NavTab.Panel key="3" title="选项三" children="选项卡内容三" />
              </NavTab>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
  {`<NavTab theme="line">
  <NavTab.Panel key="1" title="选项一" children="选项卡内容一" />
  <NavTab.Panel key="2" title="选项二" children="选项卡内容二" />
  <NavTab.Panel key="3" title="选项三" children="选项卡内容三" />
</NavTab>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <ParamTable data={ this.parameters } />
        </div>
      </Page>
    );
  }
}

export default NavTabExamples;
