// one-ui/docs/examples/overview.jsx
import React from 'react';
import moment from 'moment';
import Page from '../components/page/index.jsx';
import { Dropdown, DatePicker, CopyClipboard, ControlBox, Select, Modal, SearchInput, Table, Loading, Layout, NavTab } from '../../build/components';
import { peoples, columns } from '../../config/example.js';

class Overview extends React.Component {
  _microtime = undefined;

  state = {
    autoHide: false,
    microtime: undefined
  };

  handleMountModal() {
    Modal.mount({
      title: '测试模态框',
      draggable: false
    });
  }

  handleApplySelectedTime() {
    this.setState({
      microtime: this._microtime
    });

    this.handleCloseTimePicker();
  }

  handleCloseTimePicker() {
    this.setState({
      autoHide: true
    });
  }

  render() {
    return (
      <Page alias="/" right={(
        <Dropdown
          label={ this.state.microtime ? moment(this.state.microtime).format('YYYY年MM月DD日-HH时mm分') : '请选择查询范围' }
          pullRight
          autoHide={ this.state.autoHide }
          className="pull-right date-picker-example"
          menuStyle={{
            padding: 10
          }}
        >
          <span className="arrow" />
          <DatePicker onChange={(microtime) => this._microtime = microtime} />
          <div className="text-right">
            <button className="btn btn-danger" onClick={ this.handleCloseTimePicker.bind(this) }>取消</button>
            <button className="btn btn-primary" onClick={ this.handleApplySelectedTime.bind(this) }>保存</button>
          </div>
        </Dropdown>
      )}>

        <div className="section">
          <h3>步骤</h3>
          <div className="example-area">
            <ul className="nav-step">
              <li className="item">幼稚园</li>
              <li className="item">小学</li>
              <li className="item">初中</li>
              <li className="item">高中</li>
              <li className="item"><a href="javascript:;" className="active">工作</a></li>
              <li className="item disabled">回农村</li>
            </ul>
          </div>
          <div className="source-code-area">
            <pre><code>
{`<ul className="nav-step">
  <li className="item">幼稚园</li>
  <li className="item">小学</li>
  <li className="item">初中</li>
  <li className="item">高中</li>
  <li className="item"><a href="javascript:;" className="active">工作</a></li>
  <li className="item disabled">回农村</li>
</ul>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>标题</h3>
          <div className="example-area">
            <h1>h1. 标题标题标题标题标题标题<small>小标题小标题</small></h1>
            <h2>h2. 标题标题标题标题标题标题<small>小标题小标题</small></h2>
            <h3>h3. 标题标题标题标题标题标题<small>小标题小标题</small></h3>
            <h4>h4. 标题标题标题标题标题标题<small>小标题小标题</small></h4>
            <h5>h5. 标题标题标题标题标题标题<small>小标题小标题</small></h5>
            <h6>h6. 标题标题标题标题标题标题<small>小标题小标题</small></h6>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<h1>h1. 标题标题标题标题标题标题<small>小标题小标题</small></h1>
<h2>h2. 标题标题标题标题标题标题<small>小标题小标题</small></h2>
<h3>h3. 标题标题标题标题标题标题<small>小标题小标题</small></h3>
<h4>h4. 标题标题标题标题标题标题<small>小标题小标题</small></h4>
<h5>h5. 标题标题标题标题标题标题<small>小标题小标题</small></h5>
<h6>h6. 标题标题标题标题标题标题<small>小标题小标题</small></h6>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>拷贝文本</h3>
          <div className="example-area">
            <CopyClipboard text="这是刚才复制的文字">
              <span style={{ cursor: 'pointer' }}>
                <button className="btn btn-success">点我复制文字</button>
              </span>
            </CopyClipboard>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<CopyClipboard text="这是刚才复制的文字">
  <a href="javascript:;" className="btn btn-success">点我复制文字</a>
</CopyClipboard>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>搜索框</h3>
          <div className="example-area">
            <SearchInput />
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="form-feedback">
  <div class="form-feedback-icon">
    <i class="icon close">&#xe64a;</i>
    <i class="icon">&#xe647;</i>
  </div>
  <input type="text" value="" class="form-control" placeholder="搜索" />
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>模态框</h3>
          <div className="example-area">
            <a href="javascript:;" className="btn btn-primary" onClick={ this.handleMountModal.bind(this) }>点我点我！</a>
          </div>
          <div className="source-code-area">
            <pre>
              <code>
{`Modal.mount({
  title: '测试模态框',
  draggable: false
});`}
              </code>
            </pre>
          </div>
        </div>

        <div className="section">
          <h3>选项卡</h3>
          <div className="example-area">
            <div className="nav-tabs-example">
              <NavTab theme="line">
                <NavTab.Panel key="1" title="选项一" children="选项卡内容一" />
                <NavTab.Panel key="2" title="选项二" children="选项卡内容二" />
                <NavTab.Panel key="3" title="选项三" children="选项卡内容三" />
              </NavTab>
            </div>
          </div>
          <div className="source-code-area">
            <pre>
              <code>
{`<NavTab theme="line">
  <NavTab.Panel key="1" title="选项一" children="选项卡内容一" />
  <NavTab.Panel key="2" title="选项二" children="选项卡内容二" />
  <NavTab.Panel key="3" title="选项三" children="选项卡内容三" />
</NavTab>`}
              </code>
            </pre>
          </div>
        </div>

        <div className="section">
          <h3>表格</h3>
          <div className="example-area">
            <ControlBox rights={[
              <span className="tool-item page-info">显示 20 of 20</span>,
              <a className="tool-item show-more" href="javascript:;">显示全部</a>,
              <div className="tool-item">
                <div className="form-feedback">
                  <input type="text" className="form-control" />
                  <i className="icon form-feedback-icon">&#xe647;</i>
                </div>
              </div>
            ]}>
              <div className="tool-item-group">
                <span className="tool-item only-icon">
                  <i className="icon">&#xe67c;</i>
                </span>
                <span className="tool-item only-icon">
                  <i className="icon">&#xe613;</i>
                </span>
                <span className="tool-item only-icon">
                  <i className="icon">&#xe628;</i>
                </span>
                <span className="tool-item only-icon active">
                  <i className="icon">&#xe65a;</i>
                </span>
              </div>
            </ControlBox>
            <Table data={ peoples } columns={ columns } />
          </div>
          <div className="source-code-area">
            <pre>
              <code>
{`<ControlBox rights={[
  <span className="tool-item page-info">显示 20 of 20</span>,
  <a className="tool-item show-more" href="javascript:;">显示全部</a>,
  <div className="tool-item">
    <div className="form-feedback">
      <input type="text" className="form-control" />
      <i className="icon form-feedback-icon">&#xe647;</i>
    </div>
  </div>
]}>
  <span className="tool-item">
    <i className="tool-icon icon">&#xe655;</i>
    <strong>筛选</strong>
  </span>
  <span className="tool-item border">
    <i className="icon">&#xe67c;</i>
  </span>
  <span className="tool-item border">
    <i className="icon">&#xe67b;</i>
  </span>
</ControlBox>

<Table data={ peoples } columns={ columns } />`}
              </code>
            </pre>
          </div>
        </div>

        <div className="section">
          <h3>表单</h3>
          <div className="example-area">
            <form action="#" className="form-example">
              <div className="form-group has-danger">
                <label htmlFor="username">username: </label>
                <div className="form-feedback">
                  <input type="text" className="form-control" id="username" />
                  <i className="icon form-feedback-icon">&#xe64a;</i>
                </div>
              </div>

              <div className="form-group has-success">
                <label htmlFor="password">password: </label>
                <div className="form-feedback">
                  <input type="text" className="form-control" id="password" />
                  <i className="icon form-feedback-icon">&#xe63e;</i>
                </div>
              </div>

              <div className="form-group has-danger">
                <a className="pull-right">忘记密码？</a>
                <div className="checkbox">
                  <label htmlFor="agree">
                    <input type="checkbox" value="1" name="agree" id="agree" /> 同意《用户协议》
                  </label>
                </div>
              </div>
              <div className="form-group">
                <input type="submit" value="注册" className="btn btn-primary btn-block" />
              </div>
            </form>
          </div>
          <div className="source-code-area">
            <pre>
              <code>
{`<form action="#" className="form-example">
  <div className="form-group has-danger">
    <label htmlFor="username">username: </label>
    <div className="form-feedback">
      <input type="text" className="form-control" id="username" />
      <i className="icon form-feedback-icon">&#xe64a;</i>
    </div>
  </div>

  <div className="form-group has-success">
    <label htmlFor="password">password: </label>
    <div className="form-feedback">
      <input type="text" className="form-control" id="password" />
      <i className="icon form-feedback-icon">&#xe63e;</i>
    </div>
  </div>

  <div className="form-group has-danger">
    <a className="pull-right">忘记密码？</a>
    <div className="checkbox">
      <label htmlFor="agree">
        <input type="checkbox" value="1" name="agree" id="agree" /> 同意《用户协议》
      </label>
    </div>
  </div>
  <div className="form-group">
    <input type="submit" value="注册" className="btn btn-primary btn-block" />
  </div>
</form>`}
              </code>
            </pre>
          </div>
        </div>

        <div className="section">
          <h3>加载器</h3>
          <div className="example-area">
            <Loading children="拼命加载中..." />
          </div>
          <div className="source-code-area">
            <pre><code>
{`<div className="loading">
  <div className="loader">
    <i className="icon load-icon">&#xe652;</i>
  </div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>空数据提示</h3>
          <div className="example-area">
            <div className="empty-data-tip">
              暂无数据
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
{`<div className="empty-data-tip">
  暂无数据
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>伸缩布局</h3>
          <div className="example-area flexible-examples">
            <Layout direction="row" grid={ false }>
              <Layout direction={ null }>
                flexbox
              </Layout>
              <Layout direction={ null }>
                flexbox
              </Layout>
              <Layout direction={ null }>
                flexbox
              </Layout>
              <Layout direction={ null }>
                flexbox
              </Layout>
            </Layout>

            <Layout grid={ false } vertical="top" className="h-50">
              align-items: top
            </Layout>

            <Layout grid={ false } vertical="middle" className="h-50">
              align-items: middle
            </Layout>

            <Layout grid={ false } vertical="bottom" className="h-50">
              align-items: bottom
            </Layout>

            <Layout grid={ false } direction="row" horizontal="left">
              <span className="flex-text-demo">justify-content: left</span>
            </Layout>

            <Layout grid={ false } direction="row" horizontal="center">
              <span className="flex-text-demo">justify-content: center</span>
            </Layout>

            <Layout grid={ false } direction="row" horizontal="right">
              <span className="flex-text-demo">justify-content: right</span>
            </Layout>

            <Layout grid={ false } direction="row" horizontal="around">
              <span className="flex-text-demo">justify-content: around</span>
              <span className="flex-text-demo">justify-content: around</span>
            </Layout>

            <Layout grid={ false } direction="row" horizontal="between">
              <span className="flex-text-demo">justify-content: between</span>
              <span className="flex-text-demo">justify-content: between</span>
            </Layout>
          </div>
          <div className="source-code-area">
            <pre><code>
{`<Layout direction="row" grid={ false }>
  <Layout direction={ null }>
    flexbox
  </Layout>
  <Layout direction={ null }>
    flexbox
  </Layout>
  <Layout direction={ null }>
    flexbox
  </Layout>
  <Layout direction={ null }>
    flexbox
  </Layout>
</Layout>

<Layout grid={ false } vertical="top" className="h-50">
  align-items: top
</Layout>

<Layout grid={ false } vertical="middle" className="h-50">
  align-items: middle
</Layout>

<Layout grid={ false } vertical="bottom" className="h-50">
  align-items: bottom
</Layout>

<Layout grid={ false } direction="row" horizontal="left">
  <span className="flex-text-demo">justify-content: left</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="center">
  <span className="flex-text-demo">justify-content: center</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="right">
  <span className="flex-text-demo">justify-content: right</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="around">
  <span className="flex-text-demo">justify-content: around</span>
  <span className="flex-text-demo">justify-content: around</span>
</Layout>

<Layout grid={ false } direction="row" horizontal="between">
  <span className="flex-text-demo">justify-content: between</span>
  <span className="flex-text-demo">justify-content: between</span>
</Layout>`}
            </code></pre>
          </div>
        </div>
      </Page>
    );
  }
}

export default Overview;
