// one-ui/docs/examples/popover.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import { Popover, OverlayTrigger } from '../../build/components';
import ParamTable from '../components/param-table/index.jsx';

class PopoverExample extends React.Component {
  static defaultProps = {
    popoverClassName: 'popover-mini-reverse'
  };

  render() {
    let text = this.props.trigger === 'click' ? '点我' : '鼠标移入',
        popover = (
          <Popover className={ this.props.popoverClassName }>
            <strong>Holy guacamole!</strong> Check this info.
          </Popover>
        );

    return (
      <div className="relative">
        <OverlayTrigger container={ this } rootClose {...this.props} overlay={ popover } placement="right">
          <button className="btn btn-danger">{ text }（右）</button>
        </OverlayTrigger>

        <OverlayTrigger container={ this } rootClose {...this.props} overlay={ popover } placement="top">
          <button className="btn btn-danger">{ text }（上）</button>
        </OverlayTrigger>

        <OverlayTrigger container={ this } rootClose {...this.props} overlay={ popover } placement="bottom">
          <button className="btn btn-danger">{ text }（下）</button>
        </OverlayTrigger>

        <OverlayTrigger container={ this } rootClose {...this.props} overlay={ popover } placement="left">
          <button className="btn btn-danger">{ text }（左）</button>
        </OverlayTrigger>
      </div>
    );
  }
}

class PopoverExamples extends React.Component {
  parameters = [{
    name: 'placement',
    type: 'String',
    defaults: <code>right</code>,
    allow: <code>top, right, bottom, left</code>,
    remark: '方向'
  }, {
    name: 'positionLeft',
    type: 'String | Number',
    remark: '指定左侧距离'
  }, {
    name: 'positionRight',
    type: 'String | Number',
    remark: '指定右侧距离'
  }, {
    name: 'title',
    type: 'String',
    remark: '标题'
  }, {
    name: 'className',
    type: 'Any',
    remark: '自定义样式类'
  }, {
    name: 'style',
    type: 'Object',
    remark: '弹出层样式'
  }];

  render() {
    return (
      <Page alias="弹出框">
        <div className="section">
          <h3>静态示例</h3>
          <p>4 个可能的弹出方向：顶部、右侧、底部和左侧。</p>
          <div className="example-area">
            <div className="popover-example">
              <div className="popover top">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 上</h3>
                <div className="popover-content">弹出框内容</div>
              </div>

              <div className="popover right">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 右</h3>
                <div className="popover-content">弹出框内容</div>
              </div>

              <div className="popover bottom">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 下</h3>
                <div className="popover-content">弹出框内容</div>
              </div>

              <div className="popover left">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 左</h3>
                <div className="popover-content">弹出框内容</div>
              </div>
            </div>

            <div className="popover-example">
              <div className="popover offset right">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 右（小三角靠上）</h3>
                <div className="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
              </div>

              <div className="popover offset left">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 左（小三角靠上）</h3>
                <div className="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
              </div>
            </div>

            <div className="popover-example">
              <div className="popover popover-mini top">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini right">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini bottom">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini left">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>
            </div>

            <div className="popover-example">
              <div className="popover popover-mini-reverse top">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini-reverse right">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini-reverse bottom">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini-reverse left">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="flex-row-1">
                <div className="flex-item flex-items-center">
                  <a href="javascript:;" data-place="top" data-tip="提示文字提示文字提示文字提示文字">tooltip - top</a>
                </div>
                <div className="flex-item flex-items-center">
                  <a href="javascript:;" data-place="right" data-tip="提示文字提示文字提示文字提示文字">tooltip - right</a>
                </div>
                <div className="flex-item flex-items-center">
                  <a href="javascript:;" data-place="bottom" data-tip="提示文字提示文字提示文字提示文字">tooltip - bottom</a>
                </div>
                <div className="flex-item flex-items-center">
                  <a href="javascript:;" data-place="left" data-tip="提示文字提示文字提示文字提示文字">tooltip - left</a>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div className="section">
          <h3>动态示例</h3>
          <div className="example-area">
            <PopoverExample />
          </div>

          <div className="example-area">
            <PopoverExample trigger="click" />
          </div>

          <p>如果你是使用非组件（data-tip="xxx"）的方法，则可以添加 <code>data-max-width</code> 属性来指定 tooltip 的最大宽度。更多高级玩法请参考：<a target="_blank" href="https://react-bootstrap.github.io/components.html#popovers">{'https://react-bootstrap.github.io/components.html#popovers'}</a></p>
        </div>

        <div className="section">
          <h3>默认样式</h3>
          <div className="example-area">
            <div className="popover-example">
              <div className="popover left">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 右（小三角靠上）</h3>
                <div className="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
              </div>

              <div className="popover right">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 左（小三角靠上）</h3>
                <div className="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="popover left">
  <div class="arrow"></div>
  <h3 class="popover-title">弹出框 - 右（小三角靠上）</h3>
  <div class="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
</div>

<div class="popover right">
  <div class="arrow"></div>
  <h3 class="popover-title">弹出框 - 左（小三角靠上）</h3>
  <div class="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>小三角像上偏移</h3>
          <p>在父级添加 <code>.offset</code> 样式类，小箭头默认偏移 13 像素（从最顶上开始计算）。</p>
          <div className="example-area">
            <div className="popover-example">
              <div className="popover offset left">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 右（小三角靠上）</h3>
                <div className="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
              </div>

              <div className="popover offset right">
                <div className="arrow"></div>
                <h3 className="popover-title">弹出框 - 左（小三角靠上）</h3>
                <div className="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="popover offset left">
  <div class="arrow"></div>
  <h3 class="popover-title">弹出框 - 右（小三角靠上）</h3>
  <div class="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
</div>

<div class="popover offset right">
  <div class="arrow"></div>
  <h3 class="popover-title">弹出框 - 左（小三角靠上）</h3>
  <div class="popover-content">上、下方向暂不支持小三角靠上效果（仅左右方向支持）</div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>小尺寸默认颜色</h3>
          <p>迷你型的弹出框可以用做 tooltip，且提供两种可选颜色，但是需要注意：不要包含头部！会被直接隐藏掉！并且迷你型的弹出层不支持小三角偏移效果！</p>
          <div className="example-area">
            <div className="popover-example">
              <div className="popover popover-mini left">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini right">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="popover popover-mini left">
  <div class="arrow"></div>
  <div class="popover-content">迷你弹出框（可用作 Tooltip.）</div>
</div>

<div class="popover popover-mini right">
  <div class="arrow"></div>
  <div class="popover-content">迷你弹出框（可用作 Tooltip.）</div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>小尺寸黑色</h3>
          <p>添加 <code>.popover-mini-reverse</code> 样式类，可以实现黑色提示框的效果。</p>
          <div className="example-area">
            <div className="popover-example">
              <div className="popover popover-mini-reverse left">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>

              <div className="popover popover-mini-reverse right">
                <div className="arrow"></div>
                <div className="popover-content">迷你弹出框（可用作 Tooltip.）</div>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="popover popover-mini-reverse left">
  <div class="arrow"></div>
  <div class="popover-content">迷你弹出框（可用作 Tooltip.）</div>
</div>

<div class="popover popover-mini-reverse right">
  <div class="arrow"></div>
  <div class="popover-content">迷你弹出框（可用作 Tooltip.）</div>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <div className="example-area">
            <ParamTable data={ this.parameters } />
          </div>
        </div>
      </Page>
    );
  }
}

export default PopoverExamples;
