// one-ui/docs/examples/search.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import ParamTable from '../components/param-table/index.jsx';
import { SearchInput } from '../../src/components';

class SearchExamples extends React.Component {
  parameters = [{
    name: 'onChange',
    type: <code>Function</code>,
    allow: '--',
    isRequired: true,
    defaults: <code>{`() => {}`}</code>,
    remark: '响应用户输入'
  }, {
    name: 'width',
    type: <code>String | Number</code>,
    allow: '--',
    isRequired: false,
    defaults: <code>{`auto`}</code>,
    remark: '宽度'
  }];

  render() {
    return (
      <Page alias="搜索框">
        <div className="section">
          <div className="example-area">
            <div className="form-group">
              <div className="form-feedback">
                <input type="text" className="form-control" placeholder="输入关键词进行搜索" />
                <i className="icon form-feedback-icon">&#xe647;</i>
              </div>
            </div>

            <div className="form-group">
              <div className="form-feedback">
                <input type="text" className="form-control input-xs" placeholder="输入关键词进行搜索" />
                <i className="icon form-feedback-icon">&#xe647;</i>
              </div>
            </div>

            <div className="form-group">
              <div className="form-feedback nav-search">
                <input type="text" className="form-control" placeholder="输入关键词进行搜索" />
                <i className="icon form-feedback-icon">&#xe647;</i>
              </div>
            </div>
          </div>

          <div className="source-code-area">
            <pre><code>
{`<div class="form-feedback">
  <input type="text" class="form-control" placeholder="输入关键词进行搜索" />
  <i class="icon form-feedback-icon">&#xe647;</i>
</div>

<div className="form-feedback nav-search">
  <input type="text" className="form-control" placeholder="输入关键词进行搜索" />
  <i className="icon form-feedback-icon">&#xe647;</i>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <ParamTable data={ this.parameters } />
        </div>

        <div className="section">
          <h3>动态示例</h3>
          <div className="example-area">
            <SearchInput />
          </div>

          <div className="source-code-area">
            <pre><code>
{`<SearchInput />`}
            </code></pre>
          </div>
        </div>
      </Page>
    );
  }
}

export default SearchExamples;
