// one-ui/docs/examples/trigger.jsx
import React from 'react';
import Page from '../components/page/index.jsx';
import SwitchButton from 'switch-button';
import ParamTable from '../components/param-table/index.jsx';

class TriggerExamples extends React.Component {
  parameters = [{
    name: 'onChange',
    type: 'Function',
    defaults: <code>function () {}</code>,
    remark: '改变状态时触发'
  }, {
    name: 'checked',
    type: 'Boolean',
    defaults: <code>false</code>,
    remark: '默认是否选中'
  }, {
    name: 'disabled',
    type: 'Boolean',
    defaults: <code>false</code>,
    remark: '默认被禁用'
  }];

  render() {
    return (
      <Page alias="开关">
        <div className="section">
          <h3>静态示例</h3>
          <div className="example-area">
            <div className="trigger">
              <span className="trigger-circle"></span>
            </div>

            <div className="trigger disabled">
              <span className="trigger-circle"></span>
            </div>

            <div className="trigger checked">
              <span className="trigger-circle"></span>
            </div>
          </div>
          <div className="source-code-area">
            <pre><code>
{`<div className="trigger">
  <span className="trigger-circle"></span>
</div>

<div className="trigger checked">
  <span className="trigger-circle"></span>
</div>`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>动态示例</h3>
          <div className="example-area">
            <SwitchButton />
            <SwitchButton checked />
            <SwitchButton disabled checked />
          </div>
          <div className="source-code-area">
            <pre><code>
{`<SwitchButton />
<SwitchButton checked />`}
            </code></pre>
          </div>
        </div>

        <div className="section">
          <h3>参数列表</h3>
          <div className="example-area">
            <ParamTable data={ this.parameters } />
          </div>
        </div>
      </Page>
    );
  }
}

export default TriggerExamples;
