// one-ui/.storybook/webpack.config.js
var path = require('path');
var I18nPlugin = require("i18n-webpack-plugin");
var imports = require('postcss-import');

const lanuages = {
  'cn': null,
  'en': require('../lang/en.json')
};

module.exports = {
  cache: true,
  devtool: 'cheap-module-source-map',
  entry: {
    'vendor': ['react', 'react-dom', 'react-router'],
    'bundle': ['babel-polyfill', './docs/entry.jsx']
  },
  context: path.join(process.cwd()),
  module: {
    loaders: [
      {
        test: /\.(js|jsx|es6)$/,
        loader: 'babel'
      },
      {
        test: /\.(scss|css)$/,
        loader: 'style!css!sass'
      },
      {
        test: /\.(woff|woff2|svg|eot|ttf|png|jpg|jpeg)(\?t=[0-9]+)?$/,
        loader: 'file',
        query: {
          context: path.join(process.cwd()),
          publicPath: '../',
          name: '[path][name]-[hash].[ext]'
        }
      }
    ]
  },
  output: {
    filename: '[name].js',
    path: path.join(__dirname, 'build'),
    publicPath: '/build/',
  },
  resolve: {
    modulesDirectories: [
      'src/components',
      'node_modules'
    ]
  },
  plugins: [
    new I18nPlugin(lanuages['cn'])
  ],
  devServer: {
    hot: true
  },
  postcss: function (webpack) {
    return [
      imports({
        addDependencyTo: webpack
      })
    ];
  }
};
