// one-ui/build/components/layout/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Layout = (_temp = _class = function (_React$Component) {
  _inherits(Layout, _React$Component);

  function Layout() {
    _classCallCheck(this, Layout);

    return _possibleConstructorReturn(this, (Layout.__proto__ || Object.getPrototypeOf(Layout)).apply(this, arguments));
  }

  _createClass(Layout, [{
    key: 'render',
    value: function render() {
      var _cx;

      var _props = this.props,
          direction = _props.direction,
          vertical = _props.vertical,
          alignSelf = _props.alignSelf,
          grid = _props.grid,
          horizontal = _props.horizontal,
          isFlex = this.props.type === 'flex';


      var classNames = '';

      if (isFlex) {
        if (Boolean(+grid) === false) {
          classNames = typeof direction === 'string' ? 'flex-' + direction : 'flex'; // flex, flex-row, flex-col
        } else {
          if (typeof +grid === 'number') {
            if (typeof direction === 'string') {
              // flex-row-* flex-col-*
              classNames = 'flex-' + direction + '-' + grid;
            } else {
              classNames = 'flex-' + grid; // flex-*
            }
          }
        }
      } else {
        classNames = 'pure-u-' + grid; // pure-u-*-*
      }

      var style = {};

      if (!_.isUndefined(this.props.height)) {
        style = isFlex ? { flexBasis: this.props.height } : { height: this.props.height };
      }

      return _react2.default.createElement(
        'div',
        _extends({ className: (0, _classnames2.default)(classNames, (_cx = {}, _defineProperty(_cx, 'flex-items-' + horizontal, isFlex && typeof horizontal !== 'undefined'), _defineProperty(_cx, 'flex-items-' + vertical, isFlex && typeof vertical !== 'undefined'), _defineProperty(_cx, 'flex-' + alignSelf, isFlex && typeof alignSelf !== 'undefined'), _cx), this.props.className), style: style }, this.props.attrs),
        this.props.children
      );
    }
  }]);

  return Layout;
}(_react2.default.Component), _class.defaultProps = {
  type: 'flex',
  grid: 1,
  height: undefined,
  direction: 'col',
  alignSelf: undefined,
  horizontal: undefined,
  vertical: undefined,
  attrs: undefined,
  className: undefined
}, _class.propTypes = {
  type: _propTypes2.default.oneOf(['flex', 'grid']),
  grid: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number, _propTypes2.default.bool]),
  alignSelf: _propTypes2.default.oneOf(['top', 'center', 'bottom']),
  direction: _propTypes2.default.oneOf(['col', 'row']),
  horizontal: _propTypes2.default.oneOf(['left', 'center', 'right', 'around', 'between']),
  vertical: _propTypes2.default.oneOf(['top', 'middle', 'bottom']),
  attrs: _propTypes2.default.any,
  className: _propTypes2.default.any
}, _temp);
exports.default = Layout;
module.exports = exports['default'];
